docs/ARCHITECTURE.md

```diff
@@ -0,0 +1,77 @@
+# Architecture
+
+## Layers
+Habitat can be split into roughly 4 layers:
+
+1. P2P Transport
+2. Node Services
+3. Control
+4. Data Sources
+5. Applications
+
+The majority of Habitat's core code is within layers two and three. Layer one relies heavilly on `libp2p`, and layer four is intended to be composed mainly of third party applications.
+
+### P2P Transport
+The peer-to-peer transport layer is implemented using `libp2p`, which provides critical features such as NAT holepunching, node discovery, relaying, address determination, encrypted message transport, and more. 
+
+The most important feature provided by `libp2p` is baked in NAT holepunching support. While this is very well supported by the library, its important to understand the limitations that are naturally imposed by NATs, and the resulting contraints placed on `libp2p`. The following are some of the major challenges that have to be considered:
+
+* Hole-punching requires a relay node to perform the synchronization step of the DCUTR protocol. Both end nodes need to be able to connect to this temporary relay, so a dialing node must know beforehand what relays a destination node is attached to. Habitat identifies relays through community configuration, but this implies that each community must have at least one relay. In the future, it could be advantageous to optionally use external nodes as relays to perform synchronization.
+* `libp2p` automatically identifies potential relay nodes by having each node ask its peers to attempt to dial it across the public internet. If these peers are able to reach the node without a relay, and if the node presents a consisten IP address (its not behind a symetric NAT), then it can be marked as reachable and become a relay. Importantly, `libp2p` requires that four distinct peers make a successful connection before a node can be marked reachable. This might make things challenging for small communities. Even if they have a publicly accessible node, if they don't have enough peers for `libp2p` to confirm the node's public status, it will be unable to act as a relay, and therefore NAT holepunching capabilities will not be available to the community. 
```

Comment:
typos: consisten -> consistent, symetric -> symmetric

would be nice to link to any external documentation from libp2p about NAT hole punching if thre is nay